File: myutils/src/main/java/pl/mareklangiewicz/myutils/MyMathUtils.java
package pl.mareklangiewicz.myutils;

import android.graphics.Color;
import android.graphics.Point;
import android.graphics.PointF;
import android.graphics.Rect;
import android.graphics.RectF;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;

import java.util.Random;

/**
 * Created by Marek Langiewicz on 15.07.15.
 *
 */
public final class MyMathUtils {

    private MyMathUtils() {
        throw new AssertionError("MyMathUtils class is noninstantiable.");
    }

    static public final Random RANDOM = new Random();


    static public int scale0d(int old, int oldMax, int newMax) {
        return (int)(((long)old) * ((long)newMax) / ((long)oldMax));
    }

    static public int scale1d(int old, int oldMin, int oldMax, int newMin, int newMax) {
        return scale0d(old - oldMin, oldMax - oldMin, newMax - newMin) + newMin;
    }

    static public @NonNull Point scale2d(@NonNull Point old, @NonNull Point oldMin, @NonNull Point oldMax, @NonNull Point newMin, @NonNull Point newMax) {
        return new Point(
                scale1d(old.x, oldMin.x, oldMax.x, newMin.x, newMax.x),
                scale1d(old.y, oldMin.y, oldMax.y, newMin.y, newMax.y));
    }

    static public @NonNull Point scale2d(@NonNull Point old, @NonNull Rect oldRange, @NonNull Rect newRange) {
        return scale2d(
                old,
                new Point(oldRange.left, oldRange.top),
                new Point(oldRange.right, oldRange.bottom),
                new Point(newRange.left, newRange.top),
                new Point(newRange.right, newRange.bottom));
    }


    static public int getRandomInt(int min, int max) {
        return scale1d(RANDOM.nextInt(Integer.MAX_VALUE >> 4), 0, Integer.MAX_VALUE >> 4, min, max);
    }

    static public @NonNull Point getRandomPoint(@NonNull Point min, @NonNull Point max) {
        return new Point(getRandomInt(min.x, max.x), getRandomInt(min.y, max.y));
    }


    static public float scale0d(float old, float oldMax, float newMax) {
        return old * newMax / oldMax;
    }

    static public float scale1d(float old, float oldMin, float oldMax, float newMin, float newMax) {
        return scale0d(old - oldMin, oldMax - oldMin, newMax - newMin) + newMin;
    }

    static public @NonNull PointF scale2d(@NonNull PointF old, @NonNull PointF oldMin, @NonNull PointF oldMax, @NonNull PointF newMin, @NonNull PointF newMax) {
        return new PointF(
                scale1d(old.x, oldMin.x, oldMax.x, newMin.x, newMax.x),
                scale1d(old.y, oldMin.y, oldMax.y, newMin.y, newMax.y));
    }

    static public @NonNull PointF scale2d(@NonNull PointF old, @NonNull RectF oldRange, @NonNull RectF newRange) {
        return scale2d(
                old,
                new PointF(oldRange.left, oldRange.top),
                new PointF(oldRange.right, oldRange.bottom),
                new PointF(newRange.left, newRange.top),
                new PointF(newRange.right, newRange.bottom));
    }


    static public float getRandomFloat(float min, float max) {
        return scale1d(RANDOM.nextFloat(), 0, 1, min, max);
    }

    static public @NonNull PointF getRandomPointF(@NonNull PointF min, @NonNull PointF max) {
        return new PointF(getRandomFloat(min.x, max.x), getRandomFloat(min.y, max.y));
    }

    public static @ColorInt int getRandomColor(@ColorInt int colormin, @ColorInt int colormax) {
        return Color.argb(
                getRandomInt(Color.alpha(colormin), Color.alpha(colormax)),
                getRandomInt(Color.red(colormin), Color.red(colormax)),
                getRandomInt(Color.green(colormin), Color.green(colormax)),
                getRandomInt(Color.blue(colormin), Color.blue(colormax))
        );

    }
}

File: myutils/src/main/java/pl/mareklangiewicz/myutils/MyMathUtils.kt
package pl.mareklangiewicz.myutils

import android.graphics.Color
import android.graphics.Point
import android.graphics.PointF
import android.graphics.Rect
import android.graphics.RectF
import android.support.annotation.ColorInt

import java.util.Random

/**
 * Created by Marek Langiewicz on 15.07.15.
 */



fun Int.scale0d(oldMax: Int, newMax: Int) = (toLong() * newMax.toLong() / oldMax.toLong()).toInt()

fun Float.scale0d(oldMax: Float, newMax: Float) = this * newMax / oldMax

fun Double.scale0d(oldMax: Double, newMax: Double) = this * newMax / oldMax

fun Int.scale1d(oldMin: Int, oldMax: Int, newMin: Int, newMax: Int) = (this - oldMin).scale0d(oldMax - oldMin, newMax - newMin) + newMin

fun Float.scale1d(oldMin: Float, oldMax: Float, newMin: Float, newMax: Float) =
        (this - oldMin).scale0d(oldMax - oldMin, newMax - newMin) + newMin

fun Double.scale1d(oldMin: Double, oldMax: Double, newMin: Double, newMax: Double) =
        (this - oldMin).scale0d(oldMax - oldMin, newMax - newMin) + newMin

fun Point.scale2d(oldMin: Point, oldMax: Point, newMin: Point, newMax: Point) = Point(
        x.scale1d(oldMin.x, oldMax.x, newMin.x, newMax.x),
        y.scale1d(oldMin.y, oldMax.y, newMin.y, newMax.y))

fun Point.scale2d(oldRange: Rect, newRange: Rect) = scale2d(
        Point(oldRange.left, oldRange.top),
        Point(oldRange.right, oldRange.bottom),
        Point(newRange.left, newRange.top),
        Point(newRange.right, newRange.bottom))

fun PointF.scale2d(oldMin: PointF, oldMax: PointF, newMin: PointF, newMax: PointF) = PointF(
        x.scale1d(oldMin.x, oldMax.x, newMin.x, newMax.x),
        y.scale1d(oldMin.y, oldMax.y, newMin.y, newMax.y))

fun PointF.scale2d(oldRange: RectF, newRange: RectF) = scale2d(
        PointF(oldRange.left, oldRange.top),
        PointF(oldRange.right, oldRange.bottom),
        PointF(newRange.left, newRange.top),
        PointF(newRange.right, newRange.bottom))



val RANDOM = Random()



// result should be in the half-open range: [min, max)
fun Random.nextInt(min: Int, max: Int) = nextInt(Int.MAX_VALUE).scale1d(0, Int.MAX_VALUE, min, max)

// result should be in the half-open range: [min, max)
fun Random.nextFloat(min: Float, max: Float) = nextFloat().scale1d(0f, 1f, min, max)

// result should be in the half-open range: [min, max)
fun Random.nextDouble(min: Double, max: Double) = nextDouble().scale1d(0.0, 1.0, min, max)

// result should be in the half-open range: [min, max)
fun Random.nextPoint(min: Point, max: Point) = Point(nextInt(min.x, max.x), nextInt(min.y, max.y))

// result should be in the half-open range: [min, max)
fun Random.nextPointF(min: PointF, max: PointF) = PointF(nextFloat(min.x, max.x), nextFloat(min.y, max.y))



// result should be in the half-open range: [min, max)
@ColorInt fun Random.nextColor(@ColorInt colormin: Int, @ColorInt colormax: Int): Int = Color.argb(
        nextInt(Color.alpha(colormin), Color.alpha(colormax)),
        nextInt(Color.red(colormin), Color.red(colormax)),
        nextInt(Color.green(colormin), Color.green(colormax)),
        nextInt(Color.blue(colormin), Color.blue(colormax)))




// common random pullees:

fun Random.ints(min: Int, max: Int): IPullee<Int>  = { nextInt(min, max) }
fun Random.floats(min: Float, max: Float): IPullee<Float> = { nextFloat(min, max) }
fun Random.doubles(min: Double, max: Double): IPullee<Double> = { nextDouble(min, max) }
fun Random.points(min: Point, max: Point): IPullee<Point> = { nextPoint(min, max) }
fun Random.pointfs(min: PointF, max: PointF): IPullee<PointF> = { nextPointF(min, max) }
fun Random.colors(@ColorInt min: Int,@ColorInt max: Int): IPullee<Int> = { nextColor(min, max) }

